### PR TITLE
Improve Camera Flights

### LINKED ISSUE #802: Camera flight path in Columbus View ends at wrong destination
In Columbus view, if the camera is zoomed out further than the default camera position it doesn't fly to the default camera position if it is given as the destination.  The position at the end of the flight remains further zoomed out than the default.

Sandcastle example:

<pre><code>require([
    'Cesium', 'dijit/form/Button'
], function(
    Cesium, Button) {
    "use strict";
    var ellipsoid = Cesium.Ellipsoid.WGS84;

    function fly() {
        //move camera to position further than default
        widget.scene.getCamera().position = new Cesium.Cartesian3(0, -38370635.91011722, 38404680.99332635);

        var maxRadii = ellipsoid.getMaximumRadius();
        //default position
        var position = new Cesium.Cartesian3(0.0, -1.0, 1.0).normalize().multiplyByScalar(5.0 * maxRadii);
        var direction = Cesium.Cartesian3.ZERO.subtract(position).normalize();
        var right = direction.cross(Cesium.Cartesian3.UNIT_Z);
        var up = right.cross(direction);
    
        var description = {
            destination: position,
            up: up,
            direction: direction
        };
    
        var flight = Cesium.CameraFlightPath.createAnimation(widget.scene.getFrameState(), description);
        widget.scene.getAnimations().add(flight);
    }
    
    function createButtons(scene) {
        new Button({
            label : "Zoom out, then fly to default position",
            onClick : function() {
                fly(scene);
            }
        }).placeAt('toolbar');
    }

    var widget = new Cesium.CesiumWidget('cesiumContainer');
    var transitioner = new Cesium.SceneTransitioner(widget.scene, ellipsoid);
    transitioner.morphToColumbusView();
    createButtons(widget.scene);
    Sandcastle.finishedLoading();
});</code></pre>

### PR DESCRIPTION
**Pull request by bagnell**
* Improved camera flights. Fixes #802, #804, #1158, #1400, #2093, #2468
* Adds easing function and max altitude options to flights.
* Removed deprecated `Camera.clone` #2763

**TODO**

- [x] Fix/update tests
- [x] Update CHANGES.md
- [x] Deprecate unused splines and submit issues for them
- [x] Update Sandcastle example for common use cases, e.g., global view, city view, etc.

CC #1060

### CODE REVIEW

**Comment by pjcozzi** [comment]
I updated the TODO list in this issue:

> Deprecate unused splines and submit issues for them

File: Source/Scene/Camera.js [hunk 1]
```diff
@@ -2299,6 +2315,8 @@ define([
      * @param {Matrix4} [options.endTransform] Transform matrix representing the reference frame the camera will be in when the flight is completed.
      * @param {Boolean} [options.convert=true] When <code>true</code>, the destination is converted to the correct coordinate system for each scene mode. When <code>false</code>, the destination is expected
      *                  to be in the correct coordinate system.
+     * @param {Number} [options.altitude] The maximum altitude at the peak of the flight.
```
Comment (pjcozzi) [inline Source/Scene/Camera.js]: Throughout Cesium, we call this `height`.  We should do the same here unless we have a compelling reason.

Comment (pjcozzi) [inline Source/Scene/Camera.js]: Or `maximumHeight`.

File: Source/Scene/Camera.js [hunk 2]
```diff
@@ -2489,6 +2517,8 @@ define([
      * @param {Camera~FlightCompleteCallback} [options.complete] The function to execute when the flight is complete.
      * @param {Camera~FlightCancelledCallback} [options.cancel] The function to execute if the flight is cancelled.
      * @param {Matrix4} [options.endTransform] Transform matrix representing the reference frame the camera will be in when the flight is completed.
+     * @param {Number} [options.altitude] The maximum altitude at the peak of the flight.
```
Comment (pjcozzi) [inline Source/Scene/Camera.js]: Same comment throughout.

**Comment by pjcozzi** [comment]
Also added this to the **TODO**:

> Update Sandcastle example for common use cases, e.g., global view, city view, etc.

**Comment by pjcozzi** [comment]
Switch to CV, then back to 3D.  The home view is too close.  It is: http://localhost:8080/Apps/CesiumViewer/index.html?view=-90%2C26.641969128869164%2C6382419.6691765785%2C360%2C-89.84575039834716%2C0

**Comment by pjcozzi** [comment]
The flight from, for example, Exton to LA feels a bit to "square" if you know what I mean.

**Comment by pjcozzi** [comment]
> The flight from, for example, Exton to LA feels a bit too "square" if you know what I mean.

When it is a bit slower, I think this is fine.

**Comment by pjcozzi** [comment]
The Cesium API and Geocoder/Home flights seem to have different durations.  Why not have the same default for both?

**Comment by pjcozzi** [comment]
To freeze when morphing:

* Run http://localhost:8080/Apps/Sandcastle/index.html?src=Camera.html&label=Showcases
* Switch to 2D
* Fly to rectangle
* Then switch to CV.

I'm pretty sure this is because the morph started before the flight finished.

**Comment by pjcozzi** [comment]
The CV flights tend to dip early then fly low to the ground.  Instead, if we dip late then zoom closer to directly in, I bet it will look better and load fewer tiles since it is not requesting all the high res tiles when moving quickly.

For example:

* Run http://localhost:8080/Apps/Sandcastle/index.html?src=Camera.html&label=Showcases
* Switch to CV
* Select "Fly to San Diego"

**Comment by pjcozzi** [comment]
Snap before flight:

* Run http://localhost:8080/Apps/Sandcastle/index.html?src=Camera.html&label=Showcases
* Select "Fly to My Location"
* Rotate the view 180 degrees so south is pointing up
* Select "Fly to My Location" (you'll need to unselect it first)
* The camera quickly snaps before starting the flight to rotate north back to up

**Comment by pjcozzi** [comment]
Let me know when this is ready for another look, and I will verify the existing issues and try this with our 3D building demos.

**Comment by bagnell** [comment]
>The Cesium API and Geocoder/Home flights seem to have different durations.

I changed them to be the same.

>The flight from, for example, Exton to LA feels a bit too "square" if you know what I mean.

They should be better now that everything has the duration computed from the distance by default.

**Comment by bagnell** [comment]
>To freeze when morphing...

There is already an issue for that: #1880. Many of the morph bugs aren't affected by this PR. The morphs have custom flights to handle the transitions.

**Comment by bagnell** [comment]
>Snap before flight...

I couldn't reproduce this one.

**Comment by pjcozzi** [comment]
>> Snap before flight...

> I couldn't reproduce this one.

I can't anymore either.  All good.

**Comment by pjcozzi** [comment]
>> The flight from, for example, Exton to LA feels a bit too "square" if you know what I mean.

> They should be better now that everything has the duration computed from the distance by default.

Although this will help terrain/imagery loading, I suspect the default is too slow for many users...or maybe it is fine for an app, but too slow for development.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look.

**Comment by pjcozzi** [comment]
I still think the slow ease-in from the home view makes it feel like the camera doesn't start moving for a few seconds.

We may be able to get away with a fast ease-in, slow ease-out across all flights.

**Comment by pjcozzi** [comment]
Similar to #802, zoom way back in Columbus view, then zoom into "Exton, PA."  The map goes off the screen.

**Comment by pjcozzi** [comment]
I can confirm this fixes: #802, #804, #1158, #1400, #2093, #2468

We should get input from a few more people, but the initial ease-in is too slow - I think it will frustrate users who will think Cesium isn't doing anything (especially, for example, in the corner case of #1400).  This is the web; users expect speed.

**Comment by bagnell** [comment]
>Similar to #802, zoom way back in Columbus view, then zoom into "Exton, PA." The map goes off the screen.

That happens for the same reason as when you're at the home view and fly to Exton. Interpolating the pitch moves the map off the screen until you get closer.

**Comment by bagnell** [comment]
>I still think the slow ease-in from the home view makes it feel like the camera doesn't start moving for a few seconds.

That happens when you are far from the earth. The farther the camera, the longer it seems to take for the flight to start. On the other hand, if you are really close to the surface, the fast ease-in looks terrible.

**Comment by pjcozzi** [comment]
>> I still think the slow ease-in from the home view makes it feel like the camera doesn't start moving for a few seconds.

> That happens when you are far from the earth. The farther the camera, the longer it seems to take for the flight to start. On the other hand, if you are really close to the surface, the fast ease-in looks terrible.

Step function (cross over height) to determine the ease-in speed?  Or make up some `f(current-height)`.

**Comment by bagnell** [comment]
@pjcozzi I changed the easing function to be faster easing out when the camera starts at a high altitude.

**Comment by pjcozzi** [comment]
Thanks, is there anything else for this PR?

Does anyone else want to look at it?

I probably won't be able to look at it again until Saturday.

**Comment by mramato** [comment]
The home flight is now WAY to long.  Can we change it so the unified defaults use the old home flight value.

**Comment by mramato** [comment]
Change the last line of the below example to `viewer.flyTo(viewer.entities);` In this branch the end view is incorrect.

1.10: http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Polygon.html&label=Geometries
branch: http://localhost:8080/Apps/Sandcastle/index.html?src=Polygon.html&label=Geometries

**Comment by mramato** [comment]
I think this fixes #2073 as well.  @bagnell can you confirm.

File: Source/Scene/CameraFlightPath.js [hunk 3]
```diff
@@ -410,136 +262,93 @@ define([
...
-                startObject : {},
-                stopObject: {},
-                duration : 0.0,
-                complete : complete,
-                cancel: cancel
-            };
+        var duration = options.duration;
+        if (!defined(duration)) {
+            duration = Math.ceil(Cartesian3.distance(camera.position, destination) / 1000000.0) + 3.0;
+            duration = Math.min(duration, 10.0);
```
Comment (mramato) [inline Source/Scene/CameraFlightPath.js]: This line seems to be the cause of the extremely slow fights that I'm seeing (10 seconds).  I would suggest changing these two lines to the below, which I feel works much better in all of the test cases I've tried.

```
            duration = Math.ceil(Cartesian3.distance(camera.position, destination) / 1000000.0) + 1.0;
            duration = Math.min(duration, 2.0);
```

**Comment by mramato** [comment]
That's all I've got for now, this is an awesome change.

**Comment by pjcozzi** [comment]
@bagnell can you reply here: https://groups.google.com/forum/#!topic/cesium-dev/90utjnigscg

I imagine this is just use `LINEAR`?

File: CHANGES.md [hunk 4]
```diff
@@ -3,7 +3,11 @@ Change Log
 
 ### 1.11 - 2015-07-01
 
+* Breaking changes
+  * Removed `Camera.clone` that was deprecated in 1.10.
 * Deprecated
+  * `CatmullRomSpline` has been deprecated and will be removed in Cesium 1.12.
```
Comment (pjcozzi) [inline CHANGES.md]: Maybe we should rethink this: https://groups.google.com/forum/#!topic/cesium-dev/MMIdPKFNR2s

Deprecating `HermiteSpline` is still OK though.

Comment (GatorScott) [inline CHANGES.md]: Might want to keep both and add others.  Depends on what direction is taken for tools in Viewer.  Also, GML uses b-spline and cubic spline which I believe are for roads.

Comment (pjcozzi) [inline CHANGES.md]: I'm fine with not deprecating these and reevaluating in the future.

**Comment by pjcozzi** [comment]
@bagnell anything left here?  Let's aim to merge this on Monday for 1.11.

**Comment by mramato** [comment]
#2468 still seems to be a problem.

**Comment by bagnell** [comment]
@pjcozzi I removed the deprecation from `CatmullRomSpline` and `HermiteSpline`. This is ready.

File: CHANGES.md (hunk 4 above)
Comment (bagnell) [inline CHANGES.md]: @GatorScott I added cubic b-splines and bezier splines a while back for model animation. They're sitting in the `splines` branch if we ever need them.

**Comment by bagnell** [comment]
@mramato #2468 is fixed again. I broke it when adding a fix for flying to the bounding sphere of the polygon Sandcastle demo.

**Comment by bagnell** [comment]
It seems that #2073 is fixed but I haven't been able to reproduce it in master either.

**Comment by mramato** [comment]
A big :+1: from me.  @pjcozzi feel free to merge if you're happy.

**Comment by pjcozzi** [comment]
Awesome, will review and merge first thing tomorrow.

**Comment by pjcozzi** [comment]
@bagnell can you still reply to Gaurav on the [forum](https://groups.google.com/forum/#!topic/cesium-dev/90utjnigscg)?

**Pull request merged**